doc/design/tensor_design.md

```diff
@@ -0,0 +1,175 @@
+# Tensor Design
+
+
+## Allocation
+
+`Allocation` is a [RAII](http://en.cppreference.com/w/cpp/language/raii) class template, which is used to handle a piece of memory. 
+
+```cpp
+// Template parametr 'Device' can be 'CpuDevice' or 'GpuDevice'
+template <typename Device>
+class Allocation {
+public:
+    Allocation();
+    Allocation(size_t size, Device device);
+
+    // Creates a non-owned allocation
+    Allocation(void* ptr, size_t size, Device device);
+
+    ~Allocation();
+    //No copying!
+    Allocation(const Allocation&) = delete;
+    //No assigning!
+    Allocation& operator=(const Allocation&) = delete;
+
+    void* ptr() const;
+    void* end() const;
```

Comment:
Member function `end()` is simply for users' convenience. 

With `end()`, we can check the validity of a tensor's `ptr_` like this:
```cpp
CHECK(ptr_ >= alloc.ptr() && ptr_ <= alloc.end());
```